### PR TITLE
Improve Create Dialog search ranking and rewrite calculation

### PR DESCRIPTION
**Pull request by stijn-h**
Searching in the Create Dialog would not turn up good results if the search query was short. This commit implements a better search ranking that uses user preferences (favorites/history) and some word features. Additionally, the performance and code quality has been improved too. Looking through the code initially, I noticed the filtering and ranking of types and the construction of the result tree were rather tightly coupled. It was even harder to understand because of the three different origins of types (cpp, script and plugin). 

I figured it was best to rewrite most of the code. **I only modified code internal to CreateDialog, without adding new behaviour other than which search result gets highlighted**. Hopefully this makes the code easier to understand and modify later.

Notable code changes:
 - Improved search ranking with various features. Bias is added to:
    - substring matches found early in a type string,
    - type length proportion to search term, favoring shorter types (more %coverage). 
    - type in favorite list.
    - type in first 5 elements of recent list.
    - (unchanged) preferred node type, which is automatically selected for Node2D/3D and Control types
 - Recent items are now stored in an ItemList (no visual change).
 - Removed results that had a parent that matched the search term to improve clarity.

Performance:
 - Reduce types to process upon opening the dialog (instead of every search change), usually reduces number of types by 4~5.
 - Clear arrays after closing dialog instead of keeping them around.
 - Various other optimizations.

To test if it works, I have added a project below, with a custom class_name script (CustomClass) and plugin types (Base, TypeA, TypeB), which you should be able to find, favorite and instantiate from the dialog. Other than functional requirements, the main thing to test is the search ranking. **Feedback on the feature weights is appreciated, because I am not sure if they are well distributed right now.**

[GDTest.zip](https://github.com/godotengine/godot/files/4860009/GDTest.zip)

### CODE REVIEW

**Comment by Xrayez** [comment]
Thanks for working on this!

I want to report the performance impact of this PR in relation to #27333 and #34205. Namely, when you have lots of global classes (`class_name`), both create dialog popup and searching are affected:
* it takes 2x slower to popup the dialog
* it takes up to 5 seconds to search a class on typing (without this PR, it takes ~0.2 seconds).

That's with `scons target=debug`.

What do you think could add this performance penalty?

I personally think it has little to do with your PR currently. Since you do some more sophisticated ranking and stuff, it just leads to more data queries (needlessly loading and parsing classes, again likely not related to this PR).

There are several PRs which could likely fix this: #39797, #27566, #38118.

In fact I can try merging #39797 into your PR locally to see if this has something to do with your PR in particular.

But with the rest, I see no errors.

Here's the test project as well (500 global scripts): 
[script-editor-slow-startup.zip](https://github.com/godotengine/godot/files/4865355/script-editor-slow-startup.zip)

**Comment by hinlopen** [comment]
Are your metrics compared to the linked PRs, or master/3.2? 

I ran this PR with your test project, and the only thing that makes the dialog slow to open seems to be checking `EditorData::script_class_is_parent(p_type, base_type))` for every script class for which VS reports ~8ms. This function contains expensive calls to `EditorData::script_class_get_base(p_type)` and `EditorData::script_class_load_script()`, which load resources. During search, this function is called again to check if it is a preferred type. The only other expensive calls seem to be `EditorNode::get_class_icon()` and `String::get_file()`.

These functions were all called before this PR too, so it would surprise me if this PR made it slower. It would be interesting to see what the results are when combining both PRs. If these scripts are going to be cached, it should be in EditorData and not in this dialog, which your PR seems to do!

**Comment by Xrayez** [comment]
Tested on `master`. #39797 was only tested in `3.2`, but should be the same for `master`. And yeah the way you describe the trace closely relates to my own profiling.

> I actually just checked, with both our changes, opening and searching is very fast

Nice, it means that your PR has no impact on performance, and as you said it only improves it, but with the large number of global scripts it just shadows the performance gains likely. I could adapt #39797 against `master`, but maybe there's something about #38118 which could already fix the issue, so that's also worth checking out.

Thanks for testing. 🙂

File: editor/create_dialog.cpp [hunk 1]
```diff
@@ -95,364 +71,297 @@ void CreateDialog::popup_create(bool p_dont_clear, bool p_replace_mode, const St
...
+	TreeItem *item = search_options->create_item(search_options_types[inherits]);
+	_customize_search_option_item(item, p_type, cpp_type);
+	search_options_types[p_type] = item;
 }
 
-void CreateDialog::select_type(const String &p_type) {
-	TreeItem *to_select;
-	if (search_options_types.has(p_type)) {
-		to_select = search_options_types[p_type];
+void CreateDialog::_customize_search_option_item(TreeItem *item, const String &p_type, const bool cpp_type) {
```
Comment (akien-mga) [inline editor/create_dialog.cpp:227]:
```suggestion
void CreateDialog::_customize_search_option_item(TreeItem *item, const String &p_type, const bool p_cpp_type) {
```

File: editor/create_dialog.cpp [hunk 2]
```diff
@@ -95,364 +71,297 @@ void CreateDialog::popup_create(bool p_dont_clear, bool p_replace_mode, const St
...
+	}
+}
 
-		if (search_box->get_text() == "") {
-			add_type(type, search_options_types, root, &to_select);
-		} else {
-			bool found = false;
-			String type2 = type;
-			bool cpp_type2 = cpp_type;
+String CreateDialog::_top_result(const Vector<String> p_candidates, const String &search_text) const {
```
Comment (akien-mga) [inline editor/create_dialog.cpp:271]:
```suggestion
String CreateDialog::_top_result(const Vector<String> p_candidates, const String &p_search_text) const {
```

File: editor/create_dialog.cpp [hunk 3]
```diff
@@ -95,364 +71,297 @@ void CreateDialog::popup_create(bool p_dont_clear, bool p_replace_mode, const St
...
-		to_select->select(0);
-		search_options->scroll_to_item(to_select);
-		favorite->set_disabled(false);
-		favorite->set_pressed(favorite_list.find(to_select->get_text(0)) != -1);
-	}
+	return score;
+}
 
-	get_ok()->set_disabled(root->get_children() == nullptr);
+void CreateDialog::_clean_up() {
```
Comment (akien-mga) [inline editor/create_dialog.cpp:315]:
```suggestion
void CreateDialog::_cleanup() {
```
`cleanup` is quite common in the Godot codebase.

File: editor/create_dialog.cpp [hunk 4]
```diff
@@ -597,50 +485,8 @@ void CreateDialog::_favorite_toggled() {
...
-}
-
-void CreateDialog::_history_selected() {
-	TreeItem *item = recent->get_selected();
-	if (!item) {
-		return;
-	}
-
-	search_box->set_text(item->get_text(0).get_slicec(' ', 0));
+void CreateDialog::_history_selected(int idx) {
```
Comment (akien-mga) [inline editor/create_dialog.cpp:488]:
```suggestion
void CreateDialog::_history_selected(int p_idx) {
```

File: editor/create_dialog.cpp [hunk 5]
```diff
@@ -652,12 +498,12 @@ void CreateDialog::_favorite_selected() {
...
 
 	search_box->set_text(item->get_text(0).get_slicec(' ', 0));
-	recent->deselect_all();
+	recent->unselect_all();
 	_update_search();
 }
 
-void CreateDialog::_history_activated() {
-	_history_selected();
+void CreateDialog::_history_activated(int idx) {
```
Comment (akien-mga) [inline editor/create_dialog.cpp:505]:
```suggestion
void CreateDialog::_history_activated(int p_idx) {
```

File: editor/create_dialog.cpp [hunk 6]
```diff
@@ -95,364 +71,297 @@ void CreateDialog::popup_create(bool p_dont_clear, bool p_replace_mode, const St
...
+		favorite->set_disabled(true);
+		help_bit->set_text("");
+		get_ok()->set_disabled(true);
+	}
 }
 
-bool CreateDialog::_is_type_prefered(const String &type) {
-	bool cpp_type = ClassDB::class_exists(type);
-	EditorData &ed = EditorNode::get_editor_data();
+void CreateDialog::_reconstruct_branch(const String &p_type, bool cpp_type) {
```
Comment (akien-mga) [inline editor/create_dialog.cpp:204]:
```suggestion
void CreateDialog::_reconstruct_branch(const String &p_type, bool p_cpp_type) {
```

File: editor/create_dialog.h [hunk 7]
```diff
@@ -35,60 +35,65 @@
...
+	void _favorite_activated();
 
 	Variant get_drag_data_fw(const Point2 &p_point, Control *p_from);
 	bool can_drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from) const;
 	void drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from);
 
-	bool _is_class_disabled_by_feature_profile(const StringName &p_class);
-	bool _is_type_prefered(const String &type);
+	bool _is_class_disabled_by_feature_profile(const StringName &p_class) const;
+	bool _is_type_preferred(const String &type) const;
```
Comment (akien-mga) [inline editor/create_dialog.h:95]:
```suggestion
	bool _is_type_preferred(const String &p_type) const;
```
(was mismatched with the `.cpp`, please check that others are consistent too)

**Review by willnationsdev — APPROVED** [review]
Wow, you really cleaned up this code a bunch. It's *way* more readable than it was before, and I like the incorporation of the recent/favorites into your score calculations. Thanks!

As for performance, I'm guessing this won't impact things much. And besides, with this PR, fine-tuning the performance of this dialog will become a lot easier for follow-up PRs anyway.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**